lib/pci.c

```diff
@@ -221,47 +221,55 @@ pci_hdr_write(vfu_ctx_t *vfu_ctx, const char *buf, size_t count, loff_t offset)
 {
     vfu_pci_config_space_t *cfg_space;
     int ret = 0;
+    size_t _count;
+    loff_t _offset = offset;
 
     assert(vfu_ctx != NULL);
     assert(buf != NULL);
 
     cfg_space = vfu_pci_get_config_space(vfu_ctx);
 
-    switch (offset) {
-    case PCI_COMMAND:
-        ret = handle_command_write(vfu_ctx, cfg_space, buf, count);
-        break;
-    case PCI_STATUS:
-        vfu_log(vfu_ctx, LOG_INFO, "write to status ignored");
-        break;
-    case PCI_INTERRUPT_PIN:
-        vfu_log(vfu_ctx, LOG_ERR, "attempt to write read-only field IPIN");
-        ret = ERROR_INT(EINVAL);
-        break;
-    case PCI_INTERRUPT_LINE:
-        cfg_space->hdr.intr.iline = buf[0];
-        vfu_log(vfu_ctx, LOG_DEBUG, "ILINE=%0x", cfg_space->hdr.intr.iline);
-        break;
-    case PCI_LATENCY_TIMER:
-        cfg_space->hdr.mlt = (uint8_t)buf[0];
-        vfu_log(vfu_ctx, LOG_INFO, "set to latency timer to %hhx",
-                cfg_space->hdr.mlt);
-        break;
-    case PCI_BASE_ADDRESS_0:
-    case PCI_BASE_ADDRESS_1:
-    case PCI_BASE_ADDRESS_2:
-    case PCI_BASE_ADDRESS_3:
-    case PCI_BASE_ADDRESS_4:
-    case PCI_BASE_ADDRESS_5:
-        pci_hdr_write_bar(vfu_ctx, BAR_INDEX(offset), buf);
-        break;
-    case PCI_ROM_ADDRESS:
-        ret = handle_erom_write(vfu_ctx, cfg_space, buf, count);
-        break;
-    default:
-        vfu_log(vfu_ctx, LOG_INFO, "PCI config write %#lx-%#lx not handled",
-                offset, offset + count);
-        ret = ERROR_INT(EINVAL);
+    while (count > 0) {
```

Comment:
If we're trying to handle more than one register at once, I think we need to do so for all registers. Right now if I write 20 bytes starting at PCI_STATUS, the whole write is ignored, even though PCI_STATUS isn't 20 bytes long.

We also already have code that splits up writes - pci_config_space_next_segment(). Can you extend *that* instead - it can call into a utility version that looks at offset and returns the size of the register when inside the standard header. Then, you don't need to track offset/count yourself, the code in this func can stay as it is.

Reply:
Yeah doing that in pci_config_space_next_segment() is a much better idea.